### PR TITLE
Workflow for authenticating with OpenID token on the backend.

### PR DESCRIPTION
**Pull request by kuries**
Related to #8.
Dependant on #15 

Flow:
1) After a successful login (password/sso), the access_token is exchanged with an Open ID token [Link](https://spec.matrix.org/latest/client-server-api/#post_matrixclientv3useruseridopenidrequest_token).
2) A POST request with the Open ID token as body is sent to the `/api/users/verify-openid` endpoint.
3) After successful verification with the user's matrix homeserver [link](https://spec.matrix.org/v1.2/server-server-api/#get_matrixfederationv1openiduserinfo), a new server session is created for the user.
4) A success message along with the session cookie is sent back to the client side.

### CODE REVIEW

**Comment by kuries** [comment]
Commits related to this PR start from here.
[Add authenticateWithOpenId.js and use it for password and sso based login](https://github.com/MatrixDoctors/matrix-cerberus/pull/16/commits/2bd06963cd09abc922fccfed71989748435e7531)

File: backend/app/api/endpoints/users.py [hunk 1]
```diff
@@ -15,10 +19,33 @@ async def message_users():
...
+
+            server_session_data = ServerSessionData(matrix_user=data["sub"])
+
+            response = JSONResponse({"message": "success"})
+            # Creating a server session with the matrix username.
+            response = fastapi_sessions.create_session(response, data=server_session_data)
+            return response
+
+    except aiohttp.ClientConnectionError as err:
+        raise HTTPException(status_code=500, detail="Internal server error")
```
Comment (Cadair) [inline backend/app/api/endpoints/users.py:42]: We should add logging, specifically for stuff like this, we are going to need to know why it died here.

File: backend/tests/main_test.py [hunk 2]
```diff
@@ -18,28 +23,33 @@ def test_message_users():
...
-    response = client.post("api/users/changeTokens", params=params)
-
-    assert response.status_code == 200
-    assert response.json() == {"message": "success"}
-
-    # fetch the 'matrix_user' field in session database
+@pytest.mark.asyncio
+async def test_user_session_flow():
+    matrix_user = "@example_user:matrix.org"
+    with aioresponses() as mock_server:
```
Comment (Cadair) [inline backend/tests/main_test.py:29]: It's probably a little easier to read (avoids the context manager indentation) if you put this in a fixture (from the aioresponses readme):

```python import pytest
from aioresponses import aioresponses

@pytest.fixture
def mock_aioresponse():
    with aioresponses() as m:
        yield m
```

File: backend/app/api/models.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+from pydantic import BaseModel
+
+
+class OpenIdInfo(BaseModel):
```
Comment (Cadair) [inline backend/app/api/models.py:4]: Can you add docstrings to all these classes saying what they are used for?

**Review by Cadair — APPROVED** [review]

File: backend/app/core/models.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+from typing import Optional
+
+from pydantic import BaseModel
+
+
+class ServerSessionData(BaseModel):
+    """Pydantic model used to validate server session data.
+    Created first during user login.
+    """
```
Comment (Cadair) [inline backend/app/core/models.py:9]: This is a critical error on the built in Cadair linter. Please reformat as:

```suggestion
    """
    Pydantic model used to validate server session data.
    Created first during user login.
    """
```

File: frontend/src/Pages/Login.js [hunk 5]
```diff
@@ -46,6 +46,8 @@ export default function Login() {
 	// Used to disable input fields when the homeserver url is invalid.
 	const [disableFields, setDisableFields] = useState(false);
 
+	const navigate = useNavigate();
```
Comment (Cadair) [inline frontend/src/Pages/Login.js:49]: react is strange and alien to me lol

File: backend/app/api/endpoints/users.py [hunk 6]
```diff
@@ -15,11 +17,27 @@ async def message_users():
...
+@router.post("/verify-openid")
+async def verify_openid(request: Request, open_id_info: OpenIdInfo):
+    matrix_homeserver = "https://" + open_id_info.matrix_server_name
+    params = {"access_token": open_id_info.access_token}
+    url = urljoin(matrix_homeserver, "/_matrix/federation/v1/openid/userinfo")
+
+    try:
+        async with http_client.session.get(url, params=params) as resp:
+            if resp.status != 200:
+                raise HTTPException(status_code=404, detail="Invalid token")
```
[the remark this answers is not in the document]
Comment (kuries) [reply backend/app/api/endpoints/users.py:29]: I've corrected it in PR #26 under commit [50d9f7a](https://github.com/MatrixDoctors/matrix-cerberus/pull/26/commits/50d9f7a8fe090ea0d6622d8debb07b1d6696195d)

https://github.com/MatrixDoctors/matrix-cerberus/blob/50d9f7a8fe090ea0d6622d8debb07b1d6696195d/backend/app/api/api.py#L46

File: backend/app/api/deps.py [hunk 7]
```diff
@@ -0,0 +1,3 @@
+from app.core.sessions import SessionCookie
```
[the remark this answers is not in the document]
Comment (kuries) [reply backend/app/api/deps.py:4]: Done!

**Pull request merged**